classes/user/DAO.inc.php

```diff
@@ -0,0 +1,344 @@
+            foreach ($rows as $row) {
+                yield $row->user_id => $this->fromRow($row, $query->includeReviewerData);
+            }
+        });
+    }
+
+    /**
+     * Get the total count of users matching the configured query
+     */
+    public function getCount(Collector $query): int
+    {
+        return $query
+            ->getQueryBuilder()
+            ->count();
+    }
+
+    /**
+     * Get a list of ids matching the configured query
+     */
+    public function getIds(Collector $query): Collection
+    {
+        return $query
+            ->getQueryBuilder()
+            ->select('u.' . $this->primaryKeyColumn)
+            ->pluck('u.' . $this->primaryKeyColumn);
+    }
+
+    /**
+     * Retrieve a user by username.
+     *
+     * @param $username string
+     * @param $allowDisabled boolean
+     *
+     * @return User?
+     */
+    public function getByUsername(string $username, bool $allowDisabled = true): ?User
+    {
+        $row = DB::table('users')
+            ->where('username', '=', $username)
+            ->when(!$allowDisabled, function ($query) {
+                return $query->where('disabled', '=', false);
+            })
+            ->get('user_id')
+            ->first();
+        return $row ? $this->get($row->user_id) : null;
+    }
+
+    /**
+     * Retrieve a user by API key.
+     *
+     * @param $allowDisabled boolean
+     *
+     * @return User?
+     */
+    public function getByApiKey(string $apiKey, bool $allowDisabled = true): ?User
+    {
+        return $this->getBySetting('apiKey', $apiKey, $allowDisabled);
+    }
+
+    /**
+     * Retrieve a user by email address.
+     *
+     * @param $allowDisabled boolean
+     *
+     * @return User?
+     */
+    public function getByEmail(string $email, bool $allowDisabled = true): ?User
+    {
+        $row = DB::table('users')
+            ->where('email', '=', $email)
+            ->when(!$allowDisabled, function ($query) {
+                return $query->where('disabled', '=', false);
+            })
+            ->get('user_id')
+            ->first();
+        return $row ? $this->get($row->user_id) : null;
+    }
+
+    /**
+     * Retrieve a user by setting.
+     *
+     * @param $settingName string
+     * @param $settingValue string
+     * @param $allowDisabled boolean
+     *
+     * @return User?
+     */
+    public function getBySetting($settingName, $settingValue, $allowDisabled = true)
+    {
+        $row = DB::table('users AS u')
+            ->join('user_settings AS us', 'u.user_id', '=', 'us.user_id')
+            ->where('us.setting_name', '=', $settingName)
+            ->where('us.setting_value', '=', $settingValue)
+            ->when(!$allowDisabled, function ($query) {
+                return $query->where('u.disabled', '=', false);
+            })
+            ->get('u.user_id AS user_id')
+            ->first();
+        return $row ? $this->get($row->user_id) : null;
+    }
```

Comment:
This seems like it duplicates the `filterBySetting` method of the collector. Can we reuse that and just take the `first()` result?

Reply:
Agreed.